module/zfs/zvol.c

```diff
@@ -412,14 +412,11 @@ zvol_set_volsize(const char *name, uint64_t volsize)
 		os = zv->zv_objset;
 	}
 
-	doi = kmem_alloc(sizeof (dmu_object_info_t), KM_SLEEP);
-
-	if ((error = dmu_object_info(os, ZVOL_OBJ, doi)) ||
-	    (error = zvol_check_volsize(volsize, doi->doi_data_block_size)))
+	if ((error = dmu_object_info(os, ZVOL_OBJ, &doi)) ||
+	    (error = zvol_check_volsize(volsize, doi.doi_data_block_size)))
 		goto out;
 
 	error = zvol_update_volsize(volsize, os);
-	kmem_free(doi, sizeof (dmu_object_info_t));
```

Comment:
Yes, the reason, to my knowledge, is to control the stack size. @behlendorf might have more info on this.